crates/sui-proxy/src/middleware.rs

```diff
@@ -106,7 +106,16 @@ where
                 .inc();
             (e.status(), msg)
         })?;
-        let mut decoder = ProtobufDecoder::new(body.reader());
+
+        // try snappy decode, this isn't required/enforced yet so we can only try
+        // any error will simply use the bytes from above, assumed to be protobufs
+        let mut s = snap::raw::Decoder::new();
+        let intermediate = match s.decompress_vec(&body) {
+            Ok(v) => Bytes::from_iter(v).reader(),
+            Err(_) => body.reader(),
+        };
```

Comment:
Do you want to look at the headers to determine this or just go for it?

Reply:
ideally using the header would be better, but i'd need to take apart the request in this middleware, `let (parts, body) = request.into_parts()`, then check the header, then try to decode.  I thought that was maybe more work than just trying to decode it, but what do you think?  I can look for the header too, but it'd still need to fallback like this.

Reply:
I think we probably do want to check the header, it wouldn't be much more work and would give us a bit more confidence when something goes wrong, where it happened and why

Reply:
> it wouldn't be much more work and would give us a bit more confidence when something goes wrong, where it happened and why

I agree the effort to check it is low, but it significantly complicates that protobuf extraction middleware type.  I'd really prefer not to add that complexity since it actually doesn't give us a different outcome: we still pass the same bytes. For example, if we can't decode to snappy, we use what we previously decoded (which is how it works today).  If it does decode successfully, it follows the same path.  There isn't an alternate outcome here, no matter what, we use the previously decoded body bytes and that is already captured with our existing metrics and logging.  I just don't think the extra complexity provides any benefit until sui-node is sending this header all the time, but at that point, we need to enforce it sooner and not attempt to decode the bytes at all. For example, we can/should enforce it in a header check like we do with other headers:

```
pub async fn expect_content_encoding<B>(
    TypedHeader(content_encoding): TypedHeader<ContentEncoding>,
    request: Request<B>,
    next: Next<B>,
) -> Result<Response, (StatusCode, &'static str)> {
    MIDDLEWARE_HEADERS.with_label_values(&["content-encoding", &format!("{:?}", content_encoding)]);
    // TODO enable when we are certain all nodes are sending us the right header value
    // if !content_encoding.contains("snappy") {
    //     return Err((StatusCode::BAD_REQUEST, "invalid content-encoding header"));
    // }
    Ok(next.run(request).await)
}
```

As a compromise to the concern though, I added this middleware check for that header to log what we get.  When we see this go to zero, we can flip the switch, so to speak, on the proxy side to require it and adjust error logging there to reflect the expectation.

Reply:
FWIW this this should be able to be done without over complicating the middleware. It also gives us the ability to have better performance and an idea of what went wrong when the request was being processed since we aren't just trying things once at a time and picking the first one that doesn't fail.

```
diff --git a/crates/sui-proxy/src/middleware.rs b/crates/sui-proxy/src/middleware.rs
index a443ae19bc..3d4f1285ee 100644
--- a/crates/sui-proxy/src/middleware.rs
+++ b/crates/sui-proxy/src/middleware.rs
@@ -113,6 +113,11 @@ where
     type Rejection = (StatusCode, String);
 
     async fn from_request(req: Request<B>, state: &S) -> Result<Self, Self::Rejection> {
+        let is_compressed_with_snappy = req
+            .headers()
+            .get(axum::http::header::CONTENT_ENCODING)
+            .map(|v| v.as_bytes() == b"snappy")
+            .unwrap_or(false);
         let body = Bytes::from_request(req, state).await.map_err(|e| {
             let msg = format!("error extracting bytes; {e}");
             error!(msg);
@@ -124,12 +129,18 @@ where
 
         // try snappy decode, this isn't required/enforced yet so we can only try
         // any error will simply use the bytes from above, assumed to be protobufs
-        let mut s = snap::raw::Decoder::new();
-        let intermediate = match s.decompress_vec(&body) {
-            Ok(v) => Bytes::from_iter(v).reader(),
-            // when content-encoding: snappy becomes required
-            // this will become a fatal decode error in this middleware
-            Err(_) => body.reader(),
+
+        let intermediate = if is_compressed_with_snappy {
+            let mut s = snap::raw::Decoder::new();
+            let decompressed = s.decompress_vec(&body).map_err(|e| {
+                (
+                    StatusCode::BAD_REQUEST,
+                    format!("invalid snappy compression: {e}"),
+                )
+            })?;
+            Bytes::from(decompressed).reader()
+        } else {
+            body.reader()
         };
 
         let mut decoder = ProtobufDecoder::new(intermediate);
```

Reply:
I reconsidered and added a header check. let me know what you think of this versions.

Reply:
ok, updated again to what you had since it was easier to read/follow.  thanks!